### ISSUE TITLE
Ammonite build for Scala native

### ISSUE THREAD

**Issue by mkotsbak** [lihaoyi/Ammonite#676]
It would be nice to have an Ammonite for Scala native, to avoid the dependency on a JVM and faster startup. What are the missing pieces to make that happen?

**Comment by piotrkwiecinski** [comment]
@mkotsbak Scala native is not there yet. It's missing multi-threading for example and some underlying infrastructure is quite buggy. In addition it supports only Scala 2.11 for now. 

In addition all Ammonite dependencies need to support Native first.
@lihaoyi also mentioned https://github.com/lihaoyi/Ammonite/issues/437 has to be before going Native.

**Comment by lihaoyi** [comment]
FWIW we do not need multithreading. Ammonite functions fine on a single thread (except for the presentation compiler, which spawns off a worker thread whether you want it or not)

The biggest blockers for Ammonite-Native would be

- Dependencies. Ammonite uses uTest, FastParse, uPickle, Coursier, and a dozen or so other things that aren't currently on Scala-Native

- Scala Compiler: I heard @densh had some progress getting it on native, but I'm guessing it'll be a while before we get it native "for realz" including macros and all that

- Execution Model: Ammonite relies on classloaders to load and execute Java Bytecode, including using the output of previous runs in subsequent runs. This is antithetical to how Scala-Native (or Scala.js for that matter) requires whole-program compilation, optimization, and dead-code elimination.

- https://github.com/lihaoyi/Ammonite/issues/437, which is required to allow Ammonite scripts to run without all the Ammonite runtime. Currently, executing scripts keeps going back forth between script-code and Ammonite-code, and we'd probably want to kill that if we want to make native-compilation possible e.g. scripts

There are probably dozens of other blockers, but these are probably the biggest ones. I'm going to close this issue since it's pretty far in the future, and it's not really an "issue" someone can work on and solve right now, but you can continue using/linking to this thread for discussions and some day when Scala-Native is more ready we can re-open it and make this happen.

**Comment by mkotsbak** [comment]
I think the dependencies should be easy to make available for Native (Shapeless is already). Maybe the filesystem library is more work, but maybe not if Native now implements the Java API.

Regarding the bytecoding issue, it should be no problem for Ammonite scripts, as they could just be complied all once and run the resulting executable.

Using a native scalac would be nice, but having the resulting executables native would be a big win for startup time after first load with compiling. So being able to cross compile to Native would be a nice first step actually.

**Comment by lihaoyi** [comment]
Yeah there are two steps we could take:

- Compiled scripts run in Scala-Native: this depends on #437 in order to work, so we can cleanly separate the compiled script execution from the Ammonite code execution

- Ammonite itself is in Scala-Native. This would require ScalaC-Native, as well as some way to make the REPL's incremental compile-run-compile-run loop work with Scala-Native's whole-program optimization and DCE

**Comment by densh** [comment]
Precompiling the scripts seems to be the way to go here. Replicating the full REPL experience on top of closed-world-optimising AOT toolchain is not an easy feat to say the least. It wasn't really designed for it.
